### PR TITLE
OCPBUGS-2727: Do not fail precondition check for UnknownUpdate

### PR DESCRIPTION
**Pull request by LalatenduMohanty**
With this change we will not fail precondition checks when a version is not in available updates. So users should be able to upgrade to any version with approriate client side overrides. For example using --allow-explicit-upgrade and --to-imge flags in "oc adm upgrade"

Signed-off-by: Lalatendu Mohanty <lmohanty@redhat.com>

### CODE REVIEW

File: pkg/payload/precondition/clusterversion/recommendedupdate.go
Comment (wking) [inline pkg/payload/precondition/clusterversion/recommendedupdate.go:107]: Should any of these `RecommendedUpdate` cases should be blocking?  Because any change we make seems like it could break somebody.

File: pkg/payload/precondition/precondition.go
Comment (wking) [inline pkg/payload/precondition/precondition.go:80]: This logic isn't quite right, because it treats the existence of a single warning error should not make the overall effect warning.  Instead, the existence of a single blocking alert should make the overall effect blocking.  And also, `force` means we won't block regardless, although it may still be worth splitting informative warnings from more serious "would have been blocking except for force" warnings in that case.

Comment (LalatenduMohanty) [reply pkg/payload/precondition/precondition.go:80]: The way the current code is written it ignores the warning if there are multiple precondition check failed. Also if there is one error and NonBlockingWarning is not set to true then it will be blocking.

File: pkg/payload/precondition/precondition.go
Comment (petr-muller) [inline pkg/payload/precondition/precondition.go:19]: nit
```suggestion
	NonBlockingWarning bool // For some errors we do not want to fail the precondition check but we want to communicate about it
```

File: pkg/payload/precondition/precondition.go
Comment (LalatenduMohanty) [reply pkg/payload/precondition/precondition.go]: https://github.com/openshift/cluster-version-operator/commit/a1e9171fb415e41ff19f32c5741eb3c5e7a16e7d

File: pkg/payload/precondition/precondition.go
Comment (LalatenduMohanty) [reply pkg/payload/precondition/precondition.go]: https://github.com/openshift/cluster-version-operator/commit/a1e9171fb415e41ff19f32c5741eb3c5e7a16e7d

**Review by petr-muller — APPROVED** [review]
/lgtm
/hold

"Unqualified" review, feel free to wait for another reviewer or unhold ;)

**Comment by LalatenduMohanty** [comment]
Squashed the commits in https://github.com/openshift/cluster-version-operator/commit/e6ee0b8a52a47cf5b59e779b246d2f4ac0b8621e

File: pkg/payload/precondition/clusterversion/recommendedupdate.go
Comment (wking) [inline pkg/payload/precondition/clusterversion/recommendedupdate.go:102]: no need for this variable, you can just hard-code `NonBlockingWarning: true` belown in the `return` line.

File: pkg/payload/precondition/precondition.go
Comment (wking) [inline pkg/payload/precondition/precondition.go:94]: I don't understand what this is about.  If there are two warning messages, won't this code make the summarized condition blocking, when we want to keep it non-blocking?

File: pkg/payload/precondition/precondition.go
Comment (wking) [inline pkg/payload/precondition/precondition.go:99]: Previously we'd only get to this log line if we had blocking failures.  Now we can have either a warning or blocking issue.  I think we want something like:

```go
if force and !isWarning {
  msg = fmt.Sprintf("Forced through blocking failures: %s", msg)
  isWarning = true
}

return !isWarning, &payload.UpdateError{
  Nested:  nil,
  Reason:  "UpgradePreconditionCheckFailed",
  Message: msg,
  Name:    "PreconditionCheck",
}
```

File: pkg/payload/precondition/precondition.go
Comment (LalatenduMohanty) [reply pkg/payload/precondition/precondition.go:94]: Good catch, I forgot to remove it.

File: pkg/payload/precondition/precondition.go
Comment (LalatenduMohanty) [reply pkg/payload/precondition/precondition.go:99]: This works too. Somehow I thought we want to keep the code around force separate than the warning, just for readability

File: pkg/payload/precondition/precondition.go
Comment (LalatenduMohanty) [reply pkg/payload/precondition/precondition.go]: Resolved in https://github.com/openshift/cluster-version-operator/commit/481bcde393f244f4e4245ebec6a43941c87626de

File: pkg/payload/precondition/clusterversion/recommendedupdate.go
Comment (LalatenduMohanty) [reply pkg/payload/precondition/clusterversion/recommendedupdate.go]: Resolved in https://github.com/openshift/cluster-version-operator/commit/481bcde393f244f4e4245ebec6a43941c87626de

File: pkg/payload/precondition/precondition.go
Comment (wking) [inline pkg/payload/precondition/precondition.go:19]: nit, this could be a godoc, instead of a non-structured comment, but :shrug:

**Review by wking — APPROVED** [review]
/lgtm

**Comment by wking** [comment]
[`KubePodNotReady` for `prometheus-k8s-1`][1] is orthogonal.

/override ci/prow/e2e-agnostic

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/856/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic/1586006561224921088

**Pull request merged**
